designs/odcr.md

```diff
@@ -0,0 +1,140 @@
+
+In addition, the NodeClass with Capacity Reservations will pin the NodeClass into all instance types and availability zones the Capacity Reservations reserved.
+
+_Note that these aren't permanent restrictions but simply narrowing down what features exist in the first iteration of supporting Capacity Reservation_ 
+
+Pros: 
+- This puts the onus of checking on the user to verify capatbility of their NodeClasses and requirements. It makes no assumption about what a Capacity Reservation can do nor implies can't do (Open to checking if Capacity is at limit prior to calling CreateFleet).
+- This forces users who wish to leverage fallback and prioritization to use NodePool weights or Capacity Reservation Group rather than relying on EC2NodeClass.
+
+Cons:
+- This implementation is overly restrictive causing it to be difficult to use. Its possible that the restrictions makes it too difficult for users to use EC2NodeClasses effectively.
+
+#### Pricing and consolidation
+
+
+##### Provisioning
+
+Pricing isn't directly(?) considered during provisioning, the logic is offloaded to CreateFleet. This will remain the same as Capacity Reservation will either be used and fails, if limit is hit, or it will fall back to on demand nodes that CreateFleet chooses based on on demand allocation strategy. 
+
+##### Consolidating Capacity Reserved Instances
+
+During consolidation pricing does matter as it affects which candidate will be [prioritized](https://github.com/kubernetes-sigs/karpenter/blob/75826eb51589e546fffb594bfefa91f3850e6c82/pkg/controllers/disruption/consolidation.go#L156). Since all capacity instances are paid ahead of time, their cost is already incurred. Users would likely want to prioritize filling their capacition
+reservation first then fall back into other instances. Because of this reserved instances should likely show up as 0 dollar pricing when we calculate the instance pricing. Since each candidate is tied to a NodePool and NodeClass, we should be able to safely override the pricing per node under a capacity reservation.
+
+##### Consolidating into Capacity Reserved Instances
+
+If we track Capacity Reservation usage, we can optimize the cluster configuration by moving non-Capacity Reserved instances into 
+Capacity Reserved instances. We would need to match the instance type, platform and availability zone prior to doing this. In addition, I am wondering if it make sense as a follow up rather than the first iteration of the implementation. I believe this 
+likely deserves to be a separate controller given the complexity of consolidation
+
+#### Labels
+
+When a node is launched against a CapacityReservation we will expose Capacity Reservation
+information as labels `karpenter.k8s.aws/capacity-reservation-id`. This is helpful for users to identify those nodes that are being used
```

Comment:
Open question: If we're tying one capacity reservation ID to an EC2NodeClass, we may not need this label at all because we can just use something like `karpenter.k8s.aws/ec2nodeclass` to know which CR is being used.

Reply:
I think this is bringing in a conflicting detail that most users would not need to be aware of:
1. If we add this label, then we are creating a 1:1 of capacity reservation to nodepool; this conflicts with the statements above about using resource groups where multiple ODCR IDs are added to the group, and the group is what is used for creating capacity, not the ODCRs directly
2. This is not a "characteristic" a workload would typically look for in terms of label selectors. Workloads will look to target things such as OS platform, device manufacturer or model, etc. I would be curious to know the use cases where consumers are aware of the ODCR IDs and why they would use those for selectors over other characteristics provided by the instance details (i.e. - think NFD or GFD)

Reply:
Is there a 1 ODCR per NodePool restriction? If so, that's not clear to me. What if your CR selector terms match multiple ODCRs??

Reply:
the label is attached to the nodeclaim and Node, not Nodepool or EC2NodeClass. Its a fact that gets added to the nodeclaim when it was brought up within a capacity reservation.

Reply:
> Is there a 1 ODCR per NodePool restriction

This was a detail I either misunderstood from the original design or changed over time (I believe it's the latter). We are no longer tied 1:1 between capacity reservation and NodePool/EC2NodeClass. We can now select as many capacity reservation as we need for a given NodePool and anything that's launched in that NodePool is eligible to use _any_ of those ODCRs.